### DISCUSSION TITLE
Feature request - Module dependancy changes and module dependents.

### DISCUSSION THREAD

**Discussion by DanielWillett** [SmartlyDressedGames/Unturned-3.x-Community#4212] [General]
I'm trying to add some advanced compatibility layer stuff with one of my modules but I can't guarantee load order. It would be great to be able to specify *optional* module dependencies and dependents, to tell Unturned to load this module before dependents but after dependencies.

I understand dependants could be replaced by putting dependencies in the other module but it would be better to be able to specify like this instead because it may be owned by someone else or already released.

Example:
```jsonc
{
	"IsEnabled": true,
	"Name": "DevkitServer",
	"Version": "0.0.0.1",
        "Dependencies": [
           {
                "Name": "DevkitServer.Launcher"
                "Version": "0.0.14.0",
                // this will ensure DevkitServer.Launcher loads before this module if it exists.
                "IsOptional": true
           }
        ],
        "Dependents": [
           {
                "Name": "LevelObjectIcons"
                "Version": "1.1.1.0",
                // this will ensure LevelObjectIcons loads after this module if it exists.
                "IsOptional": true
           }
        ],
        // same as before
	"Assemblies":
	[
		{
			"Path": "/Libraries/.NET Standard 2.1/netstandard.dll",
			"Role": "Both_Optional"
		},
		{
			"Path": "/Libraries/Lib.Harmony/0Harmony.dll",
			"Role": "Both_Optional"
		},
		{
			"Path": "/Bin/DevkitServer_Server.dll",
			"Role": "Server"
		},
		{
			"Path": "/Bin/DevkitServer_Client.dll",
			"Role": "Client"
		}
	]
}
```

Only changes required as far as I can tell would be to add a little to `ModuleComparer.Compare` and add an `&& !assembly.IsOptional` to `ModuleHook.sortModules` plus a thing to check for non-optional dependents.

This would go as a great side dish to https://github.com/SmartlyDressedGames/Unturned-3.x-Community/issues/4171, just saying...

**Comment by SDGNelson** [comment]
Reasonable, but relatively low priority to improve the modules feature. Would it be reasonable for your module to load these optional add-ons itself? For example, as part of your module startup checking for the optional extras to load?

**Reply by DanielWillett** [reply]
Fair, its more for compatibility with other existing modules but I understand it's not a priority.
